File: back-end/src/main/java/service/MarcaService.java
package service;

import java.util.ArrayList;
import java.util.List;

import exception.SmartshopException;
import model.entity.Marca;
import model.entity.filtros.MarcaFiltro;
import model.repository.MarcaRepository;

public class MarcaService {

	private MarcaRepository repository = new MarcaRepository();

	public Marca salvar(Marca novaMarca) throws SmartshopException {
		if(repository.verificarMarcaExiste(novaMarca)) {
			throw new SmartshopException("Marca " + novaMarca.getNome() +" já existe, favor verificar.");
		} else {
			return repository.salvar(novaMarca);			
		}
	}

	public boolean atualizar(Marca marcaEditada) {
		return repository.alterar(marcaEditada);
	}

	public boolean excluir(int id) throws SmartshopException {
		return repository.excluir(id);
	}

	public Marca consultarPorId(int id) {
		return repository.consultarPorId(id);
	}

	public List<Marca> consultarTodas() {
		return repository.consultarTodos();
	}
	
	public ArrayList<Marca> buscarComFiltro(MarcaFiltro filtro) {
		return repository.buscarComFiltro(filtro);
	}
	
	public int contarRegistros(MarcaFiltro filtro) {
		return repository.contarRegistros(filtro);
	}
	
	public int totalPaginas(MarcaFiltro filtro) {
		return repository.totalPaginas(filtro);
	}
}

File: Frontend/src/app/shared/service/marca.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Marca } from '../model/marca';
import { Observable } from 'rxjs';
import { MarcaFiltro } from '../model/filtros/marca.filtro';

@Injectable({
  providedIn: 'root'
})
export class MarcaService {

  constructor(private httpClient: HttpClient) { }

  private readonly API: string = 'http://localhost:8080/SmartShop/rest/marca';

  salvar(marca: Marca) {
    return this.httpClient.post<Marca>(this.API, marca)
  }

  atualizar(marca: Marca): Observable<boolean> {
    return this.httpClient.put<boolean>(this.API + "/alterar", marca);
  }

  consultarTodas(): Observable<Array<Marca>> {
    return this.httpClient.get<Array<Marca>>(this.API + '/todos');
  }

  consultarComFiltro(filtro: MarcaFiltro): Observable<Array<Marca>> {
    return this.httpClient.post<Array<Marca>>(this.API + '/filtrar', filtro);
  }

  buscar(id: number): Observable<Marca> {
    return this.httpClient.get<Marca>(this.API + "/buscar/" + id);
  }

  excluir(id: number): Observable<boolean> {
    return this.httpClient.delete<boolean>(this.API + "/excluir/" + id);
  }

  contarRegistro(filtro: MarcaFiltro): Observable<number> {
    return this.httpClient.post<number>(this.API + "/contarRegistros", filtro)
  }

  contarPaginas(filtro: MarcaFiltro): Observable<number> {
    return this.httpClient.post<number>(this.API + "/totalPaginas", filtro)
  }
}
